components/AppBar/src/MDCAppBarNavigationController.m

```diff
@@ -102,23 +103,78 @@ - (void)setViewControllers:(NSArray<UIViewController *> *)viewControllers animat
 }
 
 - (void)setNavigationBarHidden:(BOOL)navigationBarHidden {
-  // TODO: Consider using this API to hide the top view controller's flexible header.
-  NSAssert(navigationBarHidden, @"%@ requires that the system navigation bar remain hidden.",
-           NSStringFromClass([self class]));
+  if (!self.shouldSetNavigationBarHiddenHideAppBar) {
+    NSAssert(navigationBarHidden, @"%@ requires that the system navigation bar remain hidden.",
+             NSStringFromClass([self class]));
+    [super setNavigationBarHidden:YES];
+    return;
+  }
 
   [super setNavigationBarHidden:YES];
+  [self appbar_setNavigationBarHidden:navigationBarHidden animated:NO];
 }
 
 - (void)setNavigationBarHidden:(BOOL)navigationBarHidden animated:(BOOL)animated {
-  // TODO: Consider using this API to hide the top view controller's flexible header.
-  NSAssert(navigationBarHidden, @"%@ requires that the system navigation bar remain hidden.",
-           NSStringFromClass([self class]));
+  if (!self.shouldSetNavigationBarHiddenHideAppBar) {
+    NSAssert(navigationBarHidden, @"%@ requires that the system navigation bar remain hidden.",
+             NSStringFromClass([self class]));
+    [super setNavigationBarHidden:YES animated:animated];
+    return;
+  }
 
   [super setNavigationBarHidden:YES animated:animated];
+  [self appbar_setNavigationBarHidden:navigationBarHidden animated:animated];
+}
+
+- (BOOL)isNavigationBarHidden {
+  if (!self.shouldSetNavigationBarHiddenHideAppBar) {
+    return [super isNavigationBarHidden];
+  }
+
+  MDCAppBarViewController *appBarViewController =
+      [self appBarViewControllerForViewController:self.visibleViewController];
+  if (!appBarViewController) {
+    return YES;
+  }
+  return appBarViewController.headerView.shiftedOffscreen;
 }
 
 #pragma mark - Private
 
+- (void)appbar_setNavigationBarHidden:(BOOL)navigationBarHidden animated:(BOOL)animated {
+  [self appbar_setNavigationBarHidden:navigationBarHidden
+                             animated:animated
+                    forViewController:self.visibleViewController];
+}
+
+- (void)appbar_setNavigationBarHidden:(BOOL)navigationBarHidden
+                             animated:(BOOL)animated
+                    forViewController:(UIViewController *)viewController {
+  MDCAppBarViewController *appBarViewController =
+      [self appBarViewControllerForViewController:viewController];
+  if (!appBarViewController) {
+    return;
+  }
+
+  if (appBarViewController.headerView.shiftBehavior == MDCFlexibleHeaderShiftBehaviorDisabled) {
+    appBarViewController.headerView.shiftBehavior = MDCFlexibleHeaderShiftBehaviorHideable;
+  }
+
+  // Only toggle visiblity on app bars that have the correct shift behavior enabled.
+  if (appBarViewController.headerView.shiftBehavior != MDCFlexibleHeaderShiftBehaviorHideable) {
```

Comment:
```
if (appBarViewController.headerView.shiftBehavior == MDCFlexibleHeaderShiftBehaviorDisabled) {
  appBarViewController.headerView.shiftBehavior = MDCFlexibleHeaderShiftBehaviorHideable;
} else if (appBarViewController.headerView.shiftBehavior != MDCFlexibleHeaderShiftBehaviorHideable) {
  return;
}
```

Reply:
I'm intentionally funneling all logical paths above through to this final gate in order to reduce the number of logical paths for completion of this method. The check here is acting in this regard as a `guard` statement.

Reply:
Yes I know, but the above snippet provides the exact same logic with needing less cycles if the shiftBehavior was Disabled. See it as a performance tip, no action needed.

Reply:
This method is not going to be on any critical path so I'd prefer to optimize for readability than performance here.

Reply:
ack